File: Stack/largest_rectangle_in_histogram.java
import java.util.Stack;

public class largest_rectangle_in_histogram {
    public static int largestRectangleArea(int[] heights) {
        /*
        Brute Force O(n^3) approach

        int largest_area = 0, height = 0;
        Stack<Integer> stack = new Stack<>();

        for (int i = 0; i < heights.length; i++) {
            height = heights[i];
            stack.push(height);
            int j = i + 1;
            while (j < heights.length && heights[i] <= heights[j]) {
                stack.push(heights[j]);
                j++;
            }
            int k = i - 1;
            while (k >= 0 && heights[i] <= heights[k]) {
                stack.push(heights[k]);
                k--;
            }

            largest_area = Math.max(largest_area, stack.size() * height);
            stack.clear();
        }

        return largest_area;

        */

        

    }
    public static void main(String[] args) {
        int[] heights = {7,1,7,2,2,4}, heights2 = {2,3};

        System.out.println(largestRectangleArea(heights));
        System.out.println(largestRectangleArea(heights2));
    }
}

File: Stack/LargestRectangleInHistogram.py
# Time Complexity: O(n)
# Space Complexity: O(n)
'''
Don't see at all how this shit is intuitive. I had to copy and paste Neetcode's soln again, but after understanding
the problem, I tried my best to explain it as short and sweet as possible. The time complexity is linear be
'''


def largestRectangleArea(heights):
        maxArea = 0
        stack = []  # pair: (index, height)

        # O(n)
        for i, h in enumerate(heights):

            # Establish the starting index 
            start = i

            # So, as long as the stack isn't empty, and the current height is
            # less then the height at the top of the stack, keep popping
            while stack and stack[-1][1] > h:

                # If indeed the current height is less then the height at the top
                # of the stack, then try to find the largest area u can with the height
                # at the top of the stack.
                index, height = stack.pop()
                maxArea = max(maxArea, height * (i - index))
                start = index
            stack.append((start, h))

        # For any remaining elements, find the largest area and see if u can find a greater one.
        # Keep in mind that for these elements, they're all able to extend all the way to the right
        # of the histogram, thats why the upper bound of the area calculation is the length of the heights array
        for i, h in stack:
            maxArea = max(maxArea, h * (len(heights) - i))
        return maxArea